coff.ml

```diff
@@ -1038,31 +1038,35 @@ end
 
 
 module Stacksize = struct
-  let set_stack_reserve filename reserve =
+  let set_stack_reserve filename size_of_stack_reserve =
     let filename =
       if not (Sys.file_exists filename) && (Sys.file_exists (filename ^ ".exe")) then filename ^ ".exe"
       else filename
     in
     let ic = open_in_bin filename in
-    let hdr_offset = int16 (read ic 0x3c 2) 0 in
-    let pe_signature = read_str ic hdr_offset 4 in
+    (* Read e_lfanew from the IMAGE_DOS_HEADER (see winnt.h) *)
+    let pe_signature_offset = int32_ (read ic 0x3c 4) 0 in
+    let pe_signature = read_str ic pe_signature_offset 4 in
     assert(pe_signature = "PE\000\000");
-    let coff_hdr = read ic 0 20 in
-    let opthdr_size = int16 coff_hdr 16 in
-    let opthdr = read ic (hdr_offset + 24) opthdr_size in
-    let machine =
-      match int16 opthdr 0 with
-      | 0x10b -> `x86
-      | 0x20b -> `x64
-      | magic -> Printf.ksprintf failwith "Cannot determine image target (magic = %x)." magic
+    let coff_file_header = read ic (pe_signature_offset + 4) 20 in
+    let size_of_optional_header = int16 coff_file_header 16 in
+    let optional_header = read ic (pe_signature_offset + 4 + 20) size_of_optional_header in
+    let pe_format =
+      match int16 optional_header 0 with
+      | 0x10b -> `PE32
+      | 0x20b -> `PE32plus
+      | magic -> Printf.ksprintf failwith "Cannot determine PE format (magic = %x)." magic
     in
-    let reserve_offset = hdr_offset + 24 + 72 in
-(*    Printf.printf "current stack reserve %ld\n%!" (int32 opthdr 72); *)
+    let size_of_stack_reserve_offset = pe_signature_offset + 4 + 20 + 72 in
+    (* begin match pe_format with
+    | `PE32 -> Printf.printf "SizeOfStackReserve 0x%08x\n%!" (int32 opthdr 72)
+    | `PE32plus -> Printf.printf "SizeOfStackReserve 0x%016Lx\n%!" (int64 opthdr 72)
+    end; *)
     close_in ic;
 
     let oc = open_out_gen [Open_wronly; Open_binary] 0x777 filename in
-    seek_out oc reserve_offset;
-    emit_int32 oc reserve;
-    if machine = `x64 then emit_int32 oc 0l;
+    seek_out oc size_of_stack_reserve_offset;
+    emit_int32 oc size_of_stack_reserve;
+    if pe_format = `PE32plus then emit_int32 oc 0l;
```

Comment:
Should we expand this to support stack sizes bigger than $2^{32}-1$ bytes?

Reply:
Unless there’s a call for it, no! I can’t find a reference to it, but I don’t think it’s supported even on 64bit Windows